File: win8_apps/javascript/Basic/Signal_Service/SignalService/js/AlljoynObjects.js
/*
 * Copyright (c) 2012, AllSeen Alliance. All rights reserved.
 *
 *    Permission to use, copy, modify, and/or distribute this software for any
 *    purpose with or without fee is hereby granted, provided that the above
 *    copyright notice and this permission notice appear in all copies.
 *
 *    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 *    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 *    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 *    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 *    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 *    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 *    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

//  Connection specifications for the bus attachment
var CONNECT_SPECS = "null:";

var APPLICATION_NAME = "signalService";

//  well-known name prefix
var WELL_KNOWN_NAME = "org.alljoyn.Bus.signal_sample";

//  Interface name for the signal service sample
var BASIC_SERVICE_INTERFACE_NAME = "org.alljoyn.Bus.signal_sample";

//  Relative path for the service 
var OBJECT_PATH = "/";

//  Session Port Num for client-service communication
var SESSION_PORT = 25;

//  Basic Service Object that will advertise the well-known interface and provide the 
//  interface to the client app
function SignalServiceBusObject(busAtt, busObjectPath) {

    this.busObject = new AllJoyn.BusObject(busAtt, busObjectPath, false);

    /* Add the interface to this object */
    var iface = null;
    var ifaceArray = new Array(iface);
    busAtt.createInterface(BASIC_SERVICE_INTERFACE_NAME, ifaceArray, false);
    ifaceArray[0].addSignal("nameChanged", "s", "newName", 0, "");
    ifaceArray[0].addProperty("name", "s", AllJoyn.PropAccessType.prop_ACCESS_RW);
    ifaceArray[0].activate();
    this.busObject.addInterface(ifaceArray[0]);
    this.serviceInterface = ifaceArray[0];

    /* This is the property value which the client will query to change */
    this.name = "";
    
    /* Called when the client calls the set property method with a string as the new
        value. Changes the property value and emits a 'nameChanged' signal */
    this.setPropertyHandler = function setPropertyHandler(event) {
        var itfName = GetEventArg(event, 0);
        var propertyName = GetEventArg(event, 1);
        var newValue = GetEventArg(event, 2);

        if (propertyName == 'name') {
            var itfName = GetEventArg(event, 0);
            var propertyName = GetEventArg(event, 1);
            var newValue = GetEventArg(event, 2).value;

            OutputLine("The client has queried to change the '" + propertyName + "' property (newName='" + newValue + "'\toldName='"+busObject.name+"')");
            busObject.name = newValue;

            try {
                /* Emit 'nameChanged' signal to anyone subscribed */
                var messageArg = new AllJoyn.MsgArg("s", new Array(newValue));
                busObject.busObject.signal("", 0, busObject.serviceInterface.getMember('nameChanged'), new Array(messageArg), 0, AllJoyn.AllJoynFlagType.alljoyn_FLAG_GLOBAL_BROADCAST);
                OutputLine("The 'nameChanged' signal was sent out as a global broadcast.");
            } catch (err) {
                OutputLine("Error occured when sending 'nameChanged' signal.");
                OutputLine(err);
            }
        }
    }
    // add and register the event and event listener from the 'set' method call
    this.busObject.addEventListener('set', this.setPropertyHandler);
    busAtt.registerBusObject(this.busObject);
}

//  This bus listener will handle bus and session events
function MyBusListener(busAtt) {

    this.sessionListener = new AllJoyn.SessionListener(busAtt);
    this.sessionPortListener = new AllJoyn.SessionPortListener(busAtt);
    this.busListener = new AllJoyn.BusListener(busAtt);

    // Called when a BusAttachment this listener is registered with is has become disconnected from the bus.
    this.BusListenerBusDisconnected = function () {
    }
    // Called when a BusAttachment this listener is registered with is stopping.
    this.BusListenerBusStopping = function () {
    }
    // Called by the bus when the listener is registered.
    this.BusListenerListenerRegistered = function (event) {
    }
    // Called by the bus when the listener is unregistered
    this.BusListenerListenerUnregistered = function () {
    }
    // Called by the bus when an external bus is discovered that is advertising a well-known name
    // that this attachment has registered interest in via a DBus call to org.alljoyn.Bus.FindAdvertisedName
    this.BusListenerFoundAdvertisedName = function (event) {
        var name = GetEventArg(event, 0);
        var transport = GetEventArg(event, 1);
        var namePrefix = GetEventArg(event, 2);
    }
    // Called by the bus when an advertisement previously reported through FoundName has become unavailable.
    this.BusListenerLostAdvertisedName = function (event) {
        var name = GetEventArg(event, 0);
        var previousOwner = GetEventArg(event, 1);
        var newOwner = GetEventArg(event, 2);
    }
    // Called by the bus when the ownership of any well-known name changes.
    this.BusListenerNameOwnerChanged = function (event) {
        var name = GetEventArg(event, 0);
        var transport = GetEventArg(event, 1);
        var namePrefix = GetEventArg(event, 2);
    };

    // Called when the session has been successfully joined.
    this.SessionPortListenerSessionJoined = function (event) {
        var sessionPort = GetEventArg(event, 0);
        if (sessionId == 0) {
            sessionId = GetEventArg(event, 1);
        }
        var joiner = GetEventArg(event, 2);
        OutputLine("A session has been established with: " + joiner + " (sessionId=" + sessionId + ").");
    }
    // Called when the a JoinSession request has been made.
    this.SessionPortListenerAcceptSessionJoiner = function (event) {
        var sessionPort = GetEventArg(event, 0);
        var sessionId = GetEventArg(event, 1);
        var joiner = GetEventArg(event, 2);
        if (sessionPort == SESSION_PORT) {
            return true;
        } else {
            return false;
        }
    }

    // Called by the bus when an existing session becomes disconnected.
    this.SessionListenerSessionLost = function (event) {
        var sessionId = GetEventArg(event, 0);
    };
    // Called by the bus when a member of a multipoint session is added.
    this.SessionListenerSessionMemberAdded = function (event) {
        var sessionId = GetEventArg(event, 0);
        var uniqueName = GetEventArg(event, 1);
    };
    // Called by the bus when a member of a multipoint session is removed.
    this.SessionListenerSessionMemberRemoved = function (event) {
        var sessionId = GetEventArg(event, 0);
        var uniqueName = GetEventArg(event, 1);
    };

    this.busListener.addEventListener('busdisconnected', this.BusListenerBusDisconnected);
    this.busListener.addEventListener('busstopping', this.BusListenerBusStopping);

    this.busListener.addEventListener('listenerregistered', this.BusListenerListenerRegistered);
    this.busListener.addEventListener('listenerunregistered', this.BusListenerListenerUnregistered);
    this.busListener.addEventListener('foundadvertisedname', this.BusListenerFoundAdvertisedName);
    this.busListener.addEventListener('lostadvertisedname', this.BusListenerLostAdvertisedName);
    this.busListener.addEventListener('nameownerchanged', this.BusListenerNameOwnerChanged);

    this.sessionPortListener.addEventListener('acceptsessionjoiner', this.SessionPortListenerAcceptSessionJoiner);
    this.sessionPortListener.addEventListener("sessionjoined", this.SessionPortListenerSessionJoined);

    this.sessionListener.addEventListener('sessionlost', this.SessionListenerSessionLost);
    this.sessionListener.addEventListener('sessionmemberadded', this.SessionListenerSessionMemberAdded);
    this.sessionListener.addEventListener('sessionmemberremoved', this.SessionListenerSessionMemberRemoved);
}

File: win8_apps/cpp/Basic/Signal_Service/SignalService/AllJoynObjects.cpp
//-----------------------------------------------------------------------
// <copyright file="AllJoynObjects.cpp" company="AllSeen Alliance.">
//     Copyright (c) 2012, AllSeen Alliance. All rights reserved.
//
//        Permission to use, copy, modify, and/or distribute this software for any
//        purpose with or without fee is hereby granted, provided that the above
//        copyright notice and this permission notice appear in all copies.
//
//        THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
//        WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
//        MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
//        ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
//        WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
//        ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
//        OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
// </copyright>
//-----------------------------------------------------------------------

#pragma once

#include "pch.h"
#include "MainPage.xaml.h"
#include "AllJoynObjects.h"
using namespace SignalService;
using namespace Windows::UI::Xaml::Controls;
using namespace Windows::UI::Xaml::Interop;
using namespace Windows::UI::Xaml;
using namespace AllJoyn;
using namespace AllJoynObjects;
using namespace Windows::Foundation;

/////////////////////////////////////Argument Object//////////////////////////////////////////////////

ArgumentObject::ArgumentObject(Platform::String ^ msg, TextBox ^ tb) : text(nullptr), textBox(nullptr)
{
    text = msg;
    textBox = tb;
}
void ArgumentObject::OnDispactched() {
    textBox->Text += text;
}


/////////////////////////////////////Basic Sample Bus Object//////////////////////////////////////////////////


BasicSampleObject::BasicSampleObject(BusAttachment ^ busAtt, Platform::String ^ path) : busObject(nullptr), name(""), signal(nullptr)
{
    busObject = ref new BusObject(busAtt, path, false);

    // Add the 'cat' interface to bus object
    InterfaceDescription ^ serviveInterface = busAtt->GetInterface(INTERFACE_NAME);
    busObject->AddInterface(serviveInterface);

    this->signal = serviveInterface->GetSignal("nameChanged");

    //	Add set handler method for when client calls setProperty
    busObject->Set += ref new BusObjectSetHandler([this]
                                                      (Platform::String ^ ifcName, Platform::String ^ propertyName, MsgArg ^ value)->QStatus
                                                  {
                                                      return SetHandler(ifcName, propertyName, value);
                                                  }
                                                  );
}

//	Called when the client calls the set property method. Set the property 'name' to
//	value specified then sends 'nameChanged' signal with new value
QStatus BasicSampleObject::SetHandler(Platform::String ^ ifcName, Platform::String ^ propertyName, MsgArg ^ value) {
    if ("name" == propertyName && ifcName == INTERFACE_NAME) {
        Platform::String ^ name = value->Value->ToString();

        App ^ app = (App ^) Application::Current;
        app->OutputLine("Set '" + propertyName + "' property was called changing name to '" + name + "'.");

        Platform::Array<Platform::Object ^, 1>^ obj = { name };
        MsgArg ^ arg = ref new MsgArg("s", obj);
        Platform::Array<MsgArg ^, 1>^ args = { arg };
        app->OutputLine("Sending the 'nameChanged' signal.\n");
        busObject->Signal("", sessionId, signal, args, 100, (unsigned char)AllJoynFlagType::ALLJOYN_FLAG_GLOBAL_BROADCAST);
    }
    return QStatus::ER_OK;
}

// return a reference to the bus object
BusObject ^ BasicSampleObject::GetBusObject::get() {
    return busObject;
}



/////////////////////////////////////My Bus Listener//////////////////////////////////////////////////

MyBusListener::MyBusListener(BusAttachment ^ busAtt) :
    busListener(nullptr), sessionPortListener(nullptr), sessionListener(nullptr)
{
    /* Create Bus Listener and register all event handlers  */
    busListener = ref new BusListener(busAtt);
    busListener->NameOwnerChanged +=
        ref new BusListenerNameOwnerChangedHandler([this]
                                                       (Platform::String ^ busName, Platform::String ^ previousOwner, Platform::String ^ newOwner)
                                                   {
                                                       NameOwnerChanged(busName, previousOwner, newOwner);
                                                   });
    busListener->BusDisconnected +=
        ref new BusListenerBusDisconnectedHandler([this] ()
                                                  {
                                                      BusDisconnected();
                                                  });
    busListener->BusStopping +=
        ref new BusListenerBusStoppingHandler([this] ()
                                              {
                                                  BusStopping();
                                              });
    busListener->ListenerRegistered +=
        ref new BusListenerListenerRegisteredHandler([this]
                                                         (BusAttachment ^ busAtt)
                                                     {
                                                         ListenerRegistered(busAtt);
                                                     });
    busListener->ListenerUnregistered +=
        ref new BusListenerListenerUnregisteredHandler([this] ()
                                                       {
                                                           ListenerUnregistered();
                                                       });
    busListener->FoundAdvertisedName +=
        ref new BusListenerFoundAdvertisedNameHandler([this]
                                                          (Platform::String ^ wellKnownName, TransportMaskType transport, Platform::String ^ namePrefix)
                                                      {
                                                          FoundAdvertisedName(wellKnownName, transport, namePrefix);
                                                      });
    busListener->LostAdvertisedName +=
        ref new BusListenerLostAdvertisedNameHandler([this]
                                                         (Platform::String ^ wellKnownName, TransportMaskType transport, Platform::String ^ namePrefix)
                                                     {
                                                         LostAdvertisedName(wellKnownName, transport, namePrefix);
                                                     });

    /* Create Session Listener and register all event handlers  */
    sessionListener = ref new SessionListener(busAtt);
    sessionListener->SessionLost +=
        ref new SessionListenerSessionLostHandler([this]
                                                      (unsigned int sessionId)
                                                  {
                                                      SessionLost(sessionId);
                                                  });
    sessionListener->SessionMemberAdded +=
        ref new SessionListenerSessionMemberAddedHandler([this]
                                                             (unsigned int sessionId, Platform::String ^ uniqueName)
                                                         {
                                                             SessionMemberAdded(sessionId, uniqueName);
                                                         });
    sessionListener->SessionMemberRemoved +=
        ref new SessionListenerSessionMemberRemovedHandler([this]
                                                               (unsigned int sessionId, Platform::String ^ uniqueName)
                                                           {
                                                               SessionMemberRemoved(sessionId, uniqueName);
                                                           });

    /* Create Session Port Listener and register all event handlers  */
    sessionPortListener = ref new SessionPortListener(busAtt);
    sessionPortListener->AcceptSessionJoiner +=
        ref new SessionPortListenerAcceptSessionJoinerHandler([this]
                                                                  (unsigned short sessionPort, Platform::String ^ joiner, SessionOpts ^ sessionOpts)->bool
                                                              {
                                                                  return AcceptSessionJoiner(sessionPort, joiner, sessionOpts);
                                                              });
    sessionPortListener->SessionJoined +=
        ref new SessionPortListenerSessionJoinedHandler([this]
                                                            (unsigned short sessionPort, unsigned int sessionId, Platform::String ^ joiner)
                                                        {
                                                            SessionJoined(sessionPort, sessionId, joiner);
                                                        });
}

// Called by the bus when an external bus is discovered that is advertising a well-known
// name that this attachment has registered interest in via a DBus call to
// org.alljoyn.Bus.FindAdvertisedName
void MyBusListener::FoundAdvertisedName(Platform::String ^ wellKnownName, TransportMaskType transport, Platform::String ^ namePrefix)
{
}
// Called by the bus when an advertisement previously reported through FoundName has become
// unavailable.
void MyBusListener::LostAdvertisedName(Platform::String ^ wellKnownName, TransportMaskType transport, Platform::String ^ namePrefix)
{
}
// Called when the owner of a well-known name changes
void MyBusListener::NameOwnerChanged(Platform::String ^ busName, Platform::String ^ previousOwner, Platform::String ^ newOwner)
{
    App ^ app = (App ^) Application::Current;
    app->OutputLine("Name Owner Changed (wkn=" + busName + " prevOwner=" + previousOwner + " newOwner=" + newOwner + ")");
}

// Called when there's been a join session request from the client
bool MyBusListener::AcceptSessionJoiner(unsigned short sessionPort, Platform::String ^ joiner, SessionOpts ^ sessionOpts)
{
    if (SERVICE_PORT == sessionPort) {
        App ^ app = (App ^) Application::Current;
        app->OutputLine("Accepting Join Session Request from joiner '" + joiner + "'.");
        return true;
    } else {
        return false;
    }
}
//	Called when a session has been joined by client
void MyBusListener::SessionJoined(unsigned short sessionPort, unsigned int sessId, Platform::String ^ joiner)
{
    App ^ app = (App ^) Application::Current;
    app->OutputLine("Join Session Success (sessionId=" + sessId + ")");
}
//	Called when a BusAttachment this listener is registered with is has become disconnected from the bus.
void MyBusListener::BusDisconnected()
{
}
//	Called when a BusAttachment this listener is registered with is stopping.
void MyBusListener::BusStopping()
{
}
//	Called by the bus when the listener is registered.
void MyBusListener::ListenerRegistered(BusAttachment ^ busAtt)
{
}
//	Called by the bus when the listener is unregistered.
void MyBusListener::ListenerUnregistered()
{
}
//	Called by the bus when an existing session becomes disconnected.
void MyBusListener::SessionLost(unsigned int sessId)
{
    App ^ app = (App ^) Application::Current;
    app->OutputLine("Session Lost (sessionId=" + sessId + ")");
}
//	Called by the bus when a member of a multipoint session is added.
void MyBusListener::SessionMemberAdded(unsigned int sessionId, Platform::String ^ uniqueName)
{
}
//	Called by the bus when a member of a multipoint session is removed.
void MyBusListener::SessionMemberRemoved(unsigned int sessionId, Platform::String ^ uniqueName)
{
}

// return a reference to the bus listener
BusListener ^ MyBusListener::GetBusListener::get() {
    return busListener;
}
// return a reference to the session listener
SessionListener ^ MyBusListener::GetSessionListener::get() {
    return sessionListener;
}
// return a reference to the session port listener
SessionPortListener ^ MyBusListener::GetSessionPortListener::get() {
    return sessionPortListener;
}
